File: python/repeated_substring_pattern.py
class Solution:
    def repeatedSubstringPattern(self, pattern: str) -> bool:
        for length in range(1, len(pattern) // 2 + 1):
            if len(pattern) % length == 0:
                substring = pattern[:length]
                if self.patternComposedOf(pattern, substring):
                    return True
        return False

    def patternComposedOf(self, pattern: str, substring: str) -> bool:
        for i in range(len(pattern) // len(substring)):
            for j in range(len(substring)):
                if substring[j] != pattern[i * len(substring) + j]:
                    return False
        return True

File: src/RepeatedSubstringPattern.java
public class RepeatedSubstringPattern {
    public static boolean repeatedSubstringPattern(String pattern) {
        for (int len = 1 ; len <= pattern.length() / 2 ; len++) {
            if (pattern.length() % len == 0) {
                String substring = pattern.substring(0, len);
                if (patternComposedOf(pattern, substring)) {
                    return true;
                }
            }
        }
        return false;
    }

    private static boolean patternComposedOf(String pattern, String subString) {
        for (int i = 0 ; i < pattern.length() / subString.length() ; i++) {
            for (int j = 0 ; j < subString.length() ; j++) {
                if (subString.charAt(j) != pattern.charAt(i * subString.length() + j)) {
                    return false;
                }
            }
        }
        return true;
    }
}
